_pandoc_filter/add_title.lua

```diff
@@ -0,0 +1,317 @@
+function fix_path (path)
+  return '.' .. path
+end
+
+
+function skip_include(el)
+
+	for i, item in ipairs(el) do
+				el[i] = pandoc.walk_block(item, {
+      Image = function (element)
+	        element.src = fix_path(element.src)
+            return element
+		   end
+		   })
+		   
+		   el[i] = pandoc.walk_block(el[i], {
+      Str = function (element)
+	        if element.text:find('{:width',1) then
+	        element.text = ("{width=")
+			end
+            return element
+		   end
+		   })
+	
+		if item == nil or item.content == nil then
+		   io.stderr:write(" ")
+		elseif item.content[1].t == nil then
+		   io.stderr:write(" ")
+		elseif #item.content >= 1 and item.content[1].t == "Str" then
+		   local str = item.content[1]
+		   if str.text:find('{%%',1) then
+			  io.stderr:write(" ")
+			  el[i] = pandoc.Para(pandoc.Str(""))
+		   elseif str.text:find('{:width',1) then
+		      io.stderr:write(" ")
+			  el[i] = pandoc.Para(pandoc.Str(""))
+		   end
+		elseif #item.content == 1 and item.content[1].t == "Image" then
+		    io.stderr:write(fix_path(item.content[1].src))
+			el[i].content[1].src = fix_path(item.content[1].src)
+	    end
+	end
+	return el 
+end
+
+function read_markdown(filename)
+	local f = io.open(filename, 'r')
+	local content = f:read('*a')
+	f:close()
+	return content
+end
+
+
+local function splitByPatternSeparator(str, sep, max)
+    sep = '^(.-)'..sep
+    local t,n,p, q,r,s = {},1,1, str:find(sep)
+    while q and n~=max do
+        t[n],n,p = s,n+1,r+1
+        q,r,s = str:find(sep,p)
+    end
+    t[n] = str:sub(p)
+    return t
+end
+
+function process_content(content, header_level)
+	local doc = pandoc.read(content)
+
+	-- now we need to create a header from the metadata
+	local title=pandoc.utils.stringify(doc.meta.title) or nil
+	_, count = content:gsub("---\n", '')
+	if count >= 2 then
+	   list = splitByPatternSeparator(content, "---\n", 3)
+	   if title == nil then
+	      content = list[3]
+	   elseif header_level == 1 then
+	      content = '# ' .. title .. '\n' .. list[3]
+	   elseif header_level == 2 then
+	      content = '## ' .. title .. '\n' .. list[3]
+	   elseif header_level == 3 then
+	      content = '### ' .. title .. '\n' .. list[3]
+	   elseif header_level == 4 then
+	      content = '#### ' .. title .. '\n' .. list[3]	
+	   end
+	   doc = pandoc.read(content)
+	else
+	   io.stderr:write("Warning: --- was not found twice at: " .. title .. "count: " .. count)
+	   local newHeader=pandoc.Header(header_level, {pandoc.Str(title)})
+	   table.insert(doc.blocks, 1, newHeader)
+	end
+    return skip_include(doc.blocks)
+
+end
+
+function filter_content(content)
```

Comment:
Is there no way to DRY this out? This jekyll has already gotten far too complex for what it is :(

That's partially my own fault :(